### PR TITLE
ESM2 Finetuning refactor

### PR DESCRIPTION
**Pull request by farhadrgh**
### Type of changes

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [x]  Refactor
- [ ]  Documentation update
- [ ]  Other (please describe):

### CI Pipeline Configuration
Configure CI behavior by checking relevant boxes below. This will automatically apply labels.

- [ ] [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [ ] [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

### Usage

```python
TODO: Add code snippet
```

### Pre-submit Checklist

 - [ ] I have tested these changes locally
 - [ ] I have updated the documentation accordingly
 - [ ] I have added/updated tests as needed
 - [ ] All existing tests pass successfully

### CODE REVIEW

**Comment by farhadrgh** [comment]
Ported over the changes from #546

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_regressor.py [hunk 1]
```diff
@@ -180,12 +180,12 @@ def get_loss_reduction_class(self) -> Type[RegressorLossReduction]:
         return RegressorLossReduction
 
 
-class InMemorySingleValueDataset(Dataset):
+class InMemorySingleValueDataset(InMemoryCSVDataset):
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_regressor.py:183]: Should we move this into `dataset.py` or anywhere under data?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_token_classifier.py [hunk 2]
```diff
@@ -205,12 +190,12 @@ def get_loss_reduction_class(self) -> Type[ClassifierLossReduction]:
         return ClassifierLossReduction
 
 
-class InMemoryPerTokenValueDataset(Dataset):
+class InMemoryPerTokenValueDataset(InMemoryCSVDataset):
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_token_classifier.py:193]: Similarly should we move this under `dataset.py`? Like `esm2/data/finetune/dataset.py` or similar.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+    )
+
+    dataset_class_options: Dict[str, Type[InMemoryCSVDataset]] = SUPPORTED_DATASETS
+
+    def dataset_class_type(desc: str) -> Type[InMemoryCSVDataset]:
+        try:
+            return dataset_class_options[desc]
+        except KeyError:
+            raise argparse.ArgumentTypeError(
+                f"Do not recognize key {desc}, valid options are: {dataset_class_options.keys()}"
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:618]: @pstjohn had a similar approach by inheriting from both `str` and `enum` to streamline argument parsing.

https://github.com/NVIDIA/bionemo-framework/blob/main/sub-packages/bionemo-llm/src/bionemo/llm/model/biobert/transformer_specs.py#L47
https://github.com/NVIDIA/bionemo-framework/blob/main/sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/train_esm2.py#L586

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+    save_top_k: int = 2,
+    nsys_profiling: bool = False,
+    nsys_start_step: int = 0,
+    nsys_end_step: Optional[int] = None,
+    nsys_ranks: List[int] = [0],
+    dataset_class: Type[InMemoryCSVDataset] = InMemorySingleValueDataset,
+    config_class: Type[BioBertConfig] = ESM2FineTuneSeqConfig,
+    metric_tracker: Callback | None = None,
+    overlap_grad_reduce: bool = True,
+    overlap_param_gather: bool = False,  # TODO waiting for a NeMo fix
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:98]: Updated on https://github.com/NVIDIA/bionemo-framework/pull/582/files

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+        metric_to_monitor_for_checkpoints=args.metric_to_monitor_for_checkpoints,
+        save_top_k=args.save_top_k,
+        nsys_profiling=args.nsys_profiling,
+        nsys_start_step=args.nsys_start_step,
+        nsys_end_step=args.nsys_end_step,
+        nsys_ranks=args.nsys_ranks,
+        dataset_class=args.dataset_class,
+        config_class=args.config_class,
+        overlap_grad_reduce=not args.no_overlap_grad_reduce,
+        overlap_param_gather=args.overlap_param_gather,
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:354]: Updated on https://github.com/NVIDIA/bionemo-framework/pull/582/files

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+    parser.add_argument(
+        "--no-overlap-grad-reduce",
+        action="store_true",
+        default=False,
+    )
+    parser.add_argument(
+        "--overlap-param-gather",
+        action="store_true",
+        default=False,
+    )  # TODO waiting for a NeMo fix
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:578]: Updated with https://github.com/NVIDIA/bionemo-framework/pull/582/files

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+
+class InMemoryCSVDataset(Dataset):
+    """An in-memory dataset that tokenize strings into BertSample instances."""
+
+    def __init__(
+        self,
+        data_path: str | os.PathLike,
+        tokenizer: tokenizer.BioNeMoESMTokenizer = tokenizer.get_tokenizer(),
+        seed: int = np.random.SeedSequence().entropy,  # type: ignore
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:47]: Would suggest a fixed integer instead. Previously we had a non-deterministic bug with golden value and turns out to be related to random random seed.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_token_classifier.py (hunk 2 above)
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_token_classifier.py:193]: Not removed yet following migration to `data.py`.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_token_classifier.py [hunk 8]
```diff
@@ -45,32 +45,17 @@
...
-    loss_mask: Tensor
-
-
-class Esm2FineTuneTokenOutput(BioBertOutput):
-    """Inference output from ESM2FineTuneTokenModel."""
-
-    classification_output: Tensor
-
-
 class ClassifierLossReduction(BERTMLMLossWithReduction):
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_token_classifier.py:51]: We can make this more general by moving to `sub-packages/bionemo-llm/src/bionemo/llm/model/loss.py`.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py [hunk 9]
```diff
@@ -0,0 +1,635 @@
...
+        metric_to_monitor_for_checkpoints=args.metric_to_monitor_for_checkpoints,
+        save_top_k=args.save_top_k,
+        nsys_profiling=args.nsys_profiling,
+        nsys_start_step=args.nsys_start_step,
+        nsys_end_step=args.nsys_end_step,
+        nsys_ranks=args.nsys_ranks,
+        dataset_class=args.dataset_class,
+        config_class=args.config_class,
+        overlap_grad_reduce=not args.no_overlap_grad_reduce,
+        overlap_param_gather=not args.overlap_param_gather,
```
Comment (sichu2023) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:358]: `not args.no_overlap_param_gather`

**Review by sichu2023 — APPROVED** [review]
LGTM!

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py (hunk 9 above)
Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:358]: Does this pass CI though?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_token_classifier.py (hunk 8 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/finetune_token_classifier.py:51]: not sure if we want to do that for a tutorial example. Will revisit that later based on the demand for generic loss fxns

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 7 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:47]: Has this been changed in pre-training datasets? is there a PR for that?

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py (hunk 9 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:358]: why wouldn't it pass?

Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:358]: Because there is no `args.overlap_param_gather`.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py (hunk 7 above)
Comment (sichu2023) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py:47]: You are right! It is still `np.random.SeedSequence().entropy`. Agree that we can keep it for consistency.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py (hunk 9 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/scripts/finetune_esm2.py:358]: hm, it passes locally, let me check and see if I am missing something

**Review by jstjohn — APPROVED** [review]

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+import torch
+import torch.utils.data
+from torch import Tensor
+from torch.utils.data import Dataset
+
+from bionemo.esm2.data import tokenizer
+from bionemo.llm.data.types import BertSample
+
+
+class InMemoryCSVDataset(Dataset):
```
[the remark this answers is not in the document]
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/dataset.py]: yes, added a classmethod in [af4a612](https://github.com/NVIDIA/bionemo-framework/pull/574/commits/af4a612e82c76104e47933f694c81e28a7cf8849)
